File: c++/day04/board.cpp
#include <array>
#include <string_view>
#include "../util.hpp"

namespace day04 {
  constexpr size_t N = 5;
  constexpr uint32_t ROW_MASK = 0b00000'00000'00000'00000'11111;
  constexpr uint32_t COL_MASK = 0b00001'00001'00001'00001'00001;

  struct Board {
    std::array<uint8_t, N * N> matrix{};
    uint32_t marked{};

    explicit Board(const std::string_view input) {
      size_t i = 0;
      for (auto& n : util::tokenize(input))
        matrix[i++] = std::stoul(n.data());
    }

    auto mark(const uint8_t num) -> void {
      for (size_t i = 0; i < matrix.size(); i += 1)
        if (matrix[i] == num)
          return void(marked |= 1 << i);
    }

    auto sum_unmarked() -> uint32_t {
      uint32_t sum = 0;
      for (size_t i = 0; i < matrix.size(); i += 1)
        if ((marked & 1 << i) == 0)
          sum += matrix[i];
      return sum;
    }

    auto bingo() const -> bool {
      for (uint32_t i = 0; i < N; i += 1) {
        const auto row = ROW_MASK << i * N;
        const auto col = COL_MASK << i;
        if ((marked & row) == row || (marked & col) == col)
          return true;
      }
      return false;
    }
  };
}

File: kotlin/day04/Board.kt
package aoc.day04

import aoc.utils.*

private const val N = 5
private const val ROW_MASK = 0b00000_00000_00000_00000_11111
private const val COL_MASK = 0b00001_00001_00001_00001_00001

internal class Board(private val matrix: List<Byte>, private var marked: Int = 0) {
  fun mark(num: Byte) {
    val shift = matrix.indexOfOrNull(num) ?: return
    marked = 1 shl shift or marked
  }

  fun hasBingo() = (0 until N).any {
    val row = ROW_MASK shl it * N
    val col = COL_MASK shl it
    marked and row == row || marked and col == col
  }

  fun sumAllUnmarked() = matrix.asSequence()
    .filterIndexed { shift, _ -> 1 shl shift and marked == 0 }
    .sum()
}

internal fun String.toBoard(): Board {
  val nums = splitToSequence(' ', '\n').filter { it.isNotEmpty() }.map { it.toByte() }
  return Board(nums.toList())
}